108: 8.1 Profiling and Reduction Counts 訳

tkyshm opened the pull request:
## 翻訳スコープ
- 8.1「Profile and Reduction Counts」

気になっている点等はインラインでコメントしていきます.

tkyshm committed:
108: 8.1 Profiling and Reduction Counts ja

tkyshm committed:
浪費と訳さず、高負荷という意味に寄せる

tkyshm committed:
日本語として意味がよく分からなかったので文を区切ってみた。

そもそも原文と意味が異なる場合があるかもしれないが、
ハイセンスの方のレビューで明確になるのを期待します。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
-\chapter{CPU and Scheduler Hogs}
+%\chapter{CPU and Scheduler Hogs}
+\chapter{CPUとスケジューラの高負荷}
```
CPU Hogsはスラングらしい. CPUリソースを食いつぶしているとか、高い負荷がかかっているような意味で用いられている様子.

hykw replied:
hog が「なんでもガツガツ食べる豚みたいな何か」のことで、hog と組み合わせの表現を結構見ます（とはいえ、cpu hog がしっくり来る訳は難しいですね。スラングによせるなら、CPUとスケジューラを食いつぶす香具師とかなんでしょうが）。

tkyshm replied:
ニュアンスを含めて訳すと、食いつぶす云々と訳すのがたしかにそうだなと思いました。
章の名前なのでシュッとした感じになるのが良いかなと思いましたので高負荷という感じに収めてみましたが、他の方の意見も聞いてみて、ベターな言い回しがあれば修正したいと思います.

shino replied:
「独占」「独り占め」という訳もあるようです。

hog - 日本語翻訳　–　Linguee辞書
https://www.linguee.jp/%E8%8B%B1%E8%AA%9E-%E6%97%A5%E6%9C%AC%E8%AA%9E/%E7%BF%BB%E8%A8%B3/hog.html

動詞的に使う場合は「CPUをガメる」とかも(ゆるい文章では)良いかもしれない。
名詞で使うなら「大食い」とか。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
-\section{Profiling and Reduction Counts}
+% While memory leaks tend to absolutely kill your system, CPU exhaustion tends to act like a bottleneck and limits the maximal work you can get out of a node. Erlang developers will have a tendency to scale horizontally when they face such issues. It is often an easy enough job to scale out the more basic pieces of code out there. Only centralized global state (process registries, ETS tables, and ...
+メモリリークはシステムを完全に停止させる傾向にありますが、CPUの枯渇はボトルネックのように振る舞う傾向があり、
+ノードから抜け出す重要な操作ができなくなることがあります。
```
後半部分、制限という言葉を直訳せずに、できなくなると訳しましたが、問題ないかどうか。
(また、意味はそもそも間違ってないか。)

hykw replied:
get out of a node で取り出されるのは性能(work)ですね。つまりCPUが枯渇するとそれがボトルネックになって、ノードから取り出せる最大性能を制限しちゃう（＝つまり、最大性能を引き出せない）という意味かと。

hykw replied:
ここの basic は何でしょうね。。。piece of code は piece of cake のもじりなんでしょうが。。。全体を見ると、そんなの基本だよ（= すごく簡単だよ）という意味のように読めますね。

ymotongpoo replied:
自分は単純に「基本的なコードが多ければ多いほどスケールが楽ちん」という解釈をしました

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+Erlang開発者は、そういった問題に直面したとき水平スケールしがちです。
+一般的に、基礎的なコードであれば、スケールアウトがとても簡単にできることが多いからです。
+ただし、集中化したグローバルの状態（プロセス辞書やETSテーブルなど）だけは、だいたい変更が必要になります。
+\footnote{通常、これはシャーディングの形を取るか、状態のレプリケーションスキームを見つけるのが適してます。最初、Erlangがきっかけで分散システムを作るが、その分散システムは大部分のプログラムのセマンティクスに適用できないと分かる。それに比べると、これらの方法はまだ適切です。}
```
この文章の訳は自信がないです。

hykw replied:
ここの basic は何でしょうね。。。piece of code は piece of cake のもじりなんでしょうが。。。全体を見ると、そんなの基本だよ（= すごく簡単だよ）という意味のように読めますね。

hykw replied:
真ん中の文は、プロセスレジストリとかETSテーブルはノード内でglobalなリソースなので、分散したらコード変更してどうにかしなきゃいけない、ということを言ってるかと。

hykw replied:
「finding a state-replication scheme that's suitable」は状態をレプリケーションする適切な方法を見つける（探す）ですかね。

ymotongpoo replied:
ここは "the form of sharding" or "the form of finding a state-repliation scheme" で、かつ that's suitable は a state-replication scheme にかかっていると思います。また and little moreは「など」程度の訳になると思います。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
 
-To pin issues to specific pieces of Erlang code, as mentioned earlier, there are two main approaches. One will be to do the old standard profiling routine, likely using one of the following applications:\footnote{All of these profilers work using Erlang tracing functionality with almost no restraint. They will have an impact on the run-time performance of the application, and shouldn't be used in ...
+% To pin issues to specific pieces of Erlang code, as mentioned earlier, there are two main approaches. One will be to do the old standard profiling routine, likely using one of the following applications:\footnote{All of these profilers work using Erlang tracing functionality with almost no restraint. They will have an impact on the run-time performance of the application, and shouldn't be used  ...
+前述したとおり、Erlangコード内の問題箇所を特定するアプローチは主に二つあります。一つは、古い標準のプロファイリングルーチンで、次に示すアプリケーションを利用することがあるでしょう。\footnote{これらのプロファイラは殆ど制限なくErlangのトレース機能を使って動作します。そのため、ランタイム上のアプリケーションのパフォーマンスに影響します。本番環境では使わないほうが良いでしょう。}
```
プロファイリングルーチン、という言葉はあるかどうか。

hykw replied:
> 一つは、古い標準のプロファイリングルーチンで、次に示すアプリケーションを利用することがあるでしょう。

routine は手順ですかね。
試訳だと「一つは、次に示すアプリケーションを利用するような、古い標準的なプロファイリングの方法です」

tkyshm replied:
手順（方法）とそのままルーチンと訳すかで悩んだので、頂いたコメントに寄せたいと思います。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
-	\item \otpapp{fprof},\footnote{\href{http://www.erlang.org/doc/man/fprof.html}{http://www.erlang.org/doc/man/fprof.html}} a more powerful replacement of eprof. It will support full concurrency and generate in-depth reports. In fact, the reports are so deep that they are usually considered opaque and hard to read. 
-	\item \otpapp{eflame},\footnote{\href{https://github.com/proger/eflame}{https://github.com/proger/eflame}} the newest kid on the block. It generates flame graphs to show deep call sequences and hot-spots in usage on a given piece of code. It allows one to quickly find issues with a single look at the final result.
+%	\item \otpapp{eprof},\footnote{\href{http://www.erlang.org/doc/man/eprof.html}{http://www.erlang.org/doc/man/eprof.html}} the oldest Erlang profiler around. It will give general percentage values and will mostly report in terms of time taken.
+ \item \otpapp{eprof}、\footnote{\href{http://www.erlang.org/doc/man/eprof.html}{http://www.erlang.org/doc/man/eprof.html}}一番古いErlangのプロファイラです。おおよそのパーセンテージ値とその所要時間を結果として出します。
```
パーセンテージだけでもよさそうかもしれない。

hykw replied:
定訳は「パーセント値」かと。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+
+%	\item \otpapp{eflame},\footnote{\href{https://github.com/proger/eflame}{https://github.com/proger/eflame}} the newest kid on the block. It generates flame graphs to show deep call sequences and hot-spots in usage on a given piece of code. It allows one to quickly find issues with a single look at the final result.
+  \item \otpapp{eflame}、\footnote{\href{https://github.com/proger/eflame}{https://github.com/proger/eflame}}
+新しいプロファイラの一つです。対象のコードの深い呼び出しシーケンスやホットスポットを可視化するFlame Graphsを生成します。
```
Flame Graphsは固有名詞かと思ったので、大文字にしてしまいましたが、適切な日本語訳があればそれにしようと思います。

hykw replied:
👍

shino replied:
本家 http://www.brendangregg.com/flamegraphs.html では、本文中、 flame graph(s) と小文字で綴っているので小文字で良いと思います。
日本語の定訳は知らないのでそのままでいいかと。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+% It is generally difficult to properly analyze the CPU usage of an Erlang node to pin problems to a specific piece of code. With everything concurrent and in a virtual machine, there is no guarantee you will find out if a specific process, driver, your own Erlang code, NIFs you may have installed, or some third-party library is eating up all your processing power.
+ErlangノードのCPU使用率を適切に分析し、問題となる特定のコードを見つけ出すのは難しいとされています。
+ある特定のプロセス、ドライバー、自前のErlangコード、組み込まれたNIF、あるいはサードパーティのライブラリーといった
+あらゆるものが並行する仮想マシンの中では、どれが処理能力を食い尽くしてるかを見つけ出せる保証はありません。
```
日本語として少しぎこちなさがあるかもと感じてますが、どうでしょうか。

ma2gedev replied:
こんな感じではどうでしょうか。

`あらゆるものが並行して仮想マシンの中にあるので、ある特定のプロセス、ドライバー、自前のErlangコード、組み込まれたNIF、あるいはサードパーティのライブラリーのどれが処理能力を食い尽くしてるかを見つけ出せる保証はありません。`

ただ元のものでもそんなに違和感はない気がします。

shino replied:
どちらも日本語としての違和感は特に感じませんでした。

一点、文の構造として、With everything concurrent and in a virtual machine, は
and の前後が対等ではないかと思います。つまり「すべてが並列であること」そして「仮想マシンのなかにあること」の両方が CPU usage だけから問題をピンポイントで当てることをむずかしくしている、と言っているのではないかと。どうでしょうか。

tkyshm commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+ノードから抜け出す重要な操作ができなくなることがあります。
+Erlang開発者は、そういった問題に直面したとき水平スケールしがちです。
+一般的に、基礎的なコードであれば、スケールアウトがとても簡単にできることが多いからです。
+ただし、集中化したグローバルの状態（プロセス辞書やETSテーブルなど）だけは、だいたい変更が必要になります。
```
僕の見聞が狭いだけかもですが、集中化という言葉は余り聞いたことないなと感じてます。ハイセンスな言葉があれば修正したいです。

hykw replied:
プロセスレジストリとかETSテーブルはノード内でglobalなリソースなので、分散したらコード変更してどうにかしなきゃいけない、ということを言ってるかと。

tkyshm replied:
ここで言いたいことは、まさにコメントのとおりなので、そのように訳したほうが分かりやすくて良いので修正します。

shino replied:
分散の妨げになる、という文脈なので中央集権(的)とかどうでしょう。

y-yu commented:
権限の問題で、僕が代理でPRタイトルの`[WIP]`をはずします。

hykw commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+% It is generally difficult to properly analyze the CPU usage of an Erlang node to pin problems to a specific piece of code. With everything concurrent and in a virtual machine, there is no guarantee you will find out if a specific process, driver, your own Erlang code, NIFs you may have installed, or some third-party library is eating up all your processing power.
+ErlangノードのCPU使用率を適切に分析し、問題となる特定のコードを見つけ出すのは難しいとされています。
+ある特定のプロセス、ドライバー、自前のErlangコード、組み込まれたNIF、あるいはサードパーティのライブラリーといった
+あらゆるものが並行する仮想マシンの中では、どれが処理能力を食い尽くしてるかを見つけ出せる保証はありません。
```
並行→並行稼動 でしょうか。

hykw commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+% It is generally difficult to properly analyze the CPU usage of an Erlang node to pin problems to a specific piece of code. With everything concurrent and in a virtual machine, there is no guarantee you will find out if a specific process, driver, your own Erlang code, NIFs you may have installed, or some third-party library is eating up all your processing power.
+ErlangノードのCPU使用率を適切に分析し、問題となる特定のコードを見つけ出すのは難しいとされています。
+ある特定のプロセス、ドライバー、自前のErlangコード、組み込まれたNIF、あるいはサードパーティのライブラリーといった
+あらゆるものが並行する仮想マシンの中では、どれが処理能力を食い尽くしてるかを見つけ出せる保証はありません。
```
並行→並行稼動 でしょうか。

tkyshm replied:
並行稼働、並行する、どちらが良さそうか、他の方レビューの意見も聞いて判断しようかと思います。

shino replied:
どちらでもこの文脈では違和感ないです。

hykw commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+あらゆるものが並行する仮想マシンの中では、どれが処理能力を食い尽くしてるかを見つけ出せる保証はありません。
+
+% The existing approaches are often limited to profiling and reduction-counting if it's in your code, and to monitoring the scheduler's work if it might be anywhere else (but also your code).
+既存のアプローチは大体限られており、コード内にある場合はプロファイリングとリダクションカウント\footnote{プロセスごとにリダクションというカウンタがあり、このカウンターは通常一つの各関数呼び出しでインクリメントされます。このリダクション回数が最大値に達した時、コンテキストスイッチが発生します。例えば、 Erlang/OTP R12B ではリダクションの最大値は2000です。}を確認する二つのアプローチ、そして、どこか他の場所（コード内も含む）にある場合はスケジューラの動作をモニタリングするアプローチがあります。
```
http://erlang.org/doc/man/erlang.html#bump_reductions-1 へのリンクがあってもいいかもしれません。

tkyshm replied:
あ、コメント付け忘れてました。ありがとうございます :pray:
参照リンクについて記載があったほうが良いかどうかというのを囲うとしてました。コメントで指摘ありましたので載せようと思います。

ma2gedev commented on 108-cpu-ja.tex:
```diff
@@ -33,11 +58,14 @@ \section{Profiling and Reduction Counts}
 
-Using the code locations\footnote{Call \expression{recon:info(PidTerm, location)} or \expression{process\_info(Pid, current\_stacktrace)} to get this information.} and current functions being run, you should be able to identify what kind of code hogs all your schedulers.
+%Using the code locations\footnote{Call \expression{recon:info(PidTerm, location)} or \expression{process\_info(Pid, current\_stacktrace)} to get this information.} and current functions being run, you should be able to identify what kind of code hogs all your schedulers.
+実行中のコードの位置\footnote{コードの位置は\expression{recon:info(PidTerm, location)}または\expression{process\_info(Pid, current\_stacktrace)}で取得してください。}や現在の関数を使って、どの種類のコードが全スケジューラをリソースを独占しているか特定できるはずです。
```
細かいですが、

> 全スケジューラをリソースを独占しているか

は `全スケジューラのリソース` か `全スケジューラリソース` でしょうか。

shino commented on 108-cpu-ja.tex:
```diff
@@ -1,24 +1,49 @@
+Erlang開発者は、そういった問題に直面したとき水平スケールしがちです。
+一般的に、基礎的なコードであれば、スケールアウトがとても簡単にできることが多いからです。
+ただし、集中化したグローバルの状態（プロセス辞書やETSテーブルなど）だけは、だいたい変更が必要になります。
+\footnote{通常、これはシャーディングの形を取るか、状態のレプリケーションスキームを見つけるのが適してます。最初、Erlangがきっかけで分散システムを作るが、その分散システムは大部分のプログラムのセマンティクスに適用できないと分かる。それに比べると、これらの方法はまだ適切です。}
```
後半難しいですね…

> It's still a decent piece of work, but nothing compared to finding out most of your program's semantics aren't applicable to distributed systems given Erlang usually forces your hand there in the first place.

ぼくの解釈した構造は:

- It's still a decent piece of work, but
- (nothing (compared to finding out most of your program's semantics) aren't applicable to
   distributed systems)
  -  (given Erlang usually forces your hand there in the first place).

ということで、ちょっと意訳かつ3文に分割してみた試訳:

これ(=finding a state-replication scheme that's suitable)は依然として相当な(decent)問題です。
しかし、あなたのプログラムのセマンティクスの大部分を理解することを除いては、分散システムの問題として
重要なこと(applicable)はなにもないのです。
なぜなら、Erlang は最初からあなたの手をそこ(=分散システム)にあるよう強制しているのですから。

これは英語できる人に見てもらいたい。 @ymotongpoo お願いします。

ymotongpoo replied:
まず given 以下は if と同じような感じであとから付け加えて訳せばいいと思うので一旦忘れます。で、構造は僕も篠原さんと同じ解釈です。ただ it の解釈が違って、僕は前の文のそのものと解釈しました。つまり状態の中央集権化の解消（shardingやfinding〜）というのは依然として大変な問題だ、というところから話が始まって、 but 以下は篠原さんと同じ解釈です。

tkyshm replied:
@shino @ymotongpoo 分かりやすいコメントありがとうございます

解釈としてはこんな感じに頭の中でまとまりました。
```
あなたのプログラムのセマンティクスの大部分を理解することに比べれば、
分散システムへの適用は大した問題にはなりません。
何故なら、最初にErlangは分散システムを強要するからです。
```

shino commented:
ワンパスレビューしました。英語難しいよ…

ymotongpoo reviewed (changes requested):
いくつかコメントしました

tkyshm committed:
hogsを大量消費と訳す

hogsは「大食い」、「リソースの大量消費」といったニュアンスを含むのでそのように修正.

tkyshm committed:
最初のパラグラフ修正.

- maximal work you get out of a node の解釈が間違っていたので修正.
- easy enough ... more ... の訳し方を修正
https://github.com/ymotongpoo/erlang-in-anger/pull/27#discussion_r198120746
のコメントの解釈が分かりやすかったのでそちらに寄せる
- 集中化 => 中央集権化, + 言葉を補完(分散システム化でコードの修正が必要)
- https://github.com/ymotongpoo/erlang-in-anger/pull/27#discussion_r197616152 以降のコメントを参考に訳を修正

tkyshm committed:
https://github.com/ymotongpoo/erlang-in-anger/pull/27#discussion_r197614240 より修正

保証されていないのは、CPUを食いつぶしてる問題をピンポイントに特定すること、というのが分かりやすいように修正.

tkyshm committed:
参照リンク追加

tkyshm committed:
ルーチン=>方法に修正

https://github.com/ymotongpoo/erlang-in-anger/pull/27#discussion_r197430564

tkyshm committed:
本家でもflame graphsと書かれているのでそちらに寄せる

tkyshm committed:
助詞がおかしかったのを修正

tkyshm committed:
fix syntax error

ymotongpoo commented on 108-cpu-ja.tex:
```diff
@@ -1,30 +1,30 @@
+基本的なコードが多ければ多いほど、スケールアウトが簡単なことが多いからです。
+中央集権的なグローバルの状態（プロセスレジストリ、ETSテーブルなど）だけは、システムの分散化に通常はコードの修正が必要です。
+\footnote{通常、これはシャーディングをするか、状態をレプリケーションする適切な方法を見つける形をとります。依然として大変な問題ではありますが、あなたのプログラムのセマンティクスの大部分を理解することに比べれば、分散システムへの適用は大した問題ではありません。何故なら、最初にErlangは分散システムを強要するからです。}
+それでも、スケールアウトする前にローカルで最適化したい場合は、CPUとスケジューラの大量消費を見つけ出だす必要があります。
```
typo:見つけ出だす

ymotongpoo commented on 108-cpu-ja.tex:
```diff
@@ -1,30 +1,30 @@
 
 % The existing approaches are often limited to profiling and reduction-counting if it's in your code, and to monitoring the scheduler's work if it might be anywhere else (but also your code).
-既存のアプローチは大体限られており、コード内にある場合はプロファイリングとリダクションカウント\footnote{プロセスごとにリダクションというカウンタがあり、このカウンターは通常一つの各関数呼び出しでインクリメントされます。このリダクション回数が最大値に達した時、コンテキストスイッチが発生します。例えば、 Erlang/OTP R12B ではリダクションの最大値は2000です。}を確認する二つのアプローチ、そして、どこか他の場所（コード内も含む）にある場合はスケジューラの動作をモニタリングするアプローチがあります。
+既存のアプローチは大体限られており、コード内にある場合はプロファイリングとリダクションカウント\footnote{プロセスごとにリダクションというカウンタがあり、このカウンターは通常一つの各関数呼び出しでインクリメントされます。このリダクション回数が最大値に達した時、コンテキストスイッチが発生します。\href{http://erlang.org/doc/man/erlang.html\#bump\_reductions-1}{http://erlang.org/doc/man/erlang.html\#bump\_reductions-1}}を確認する二つのアプローチ、そして、どこか他の場所（コード内も含む）にある場合はスケジューラの動作をモニタリングするアプローチがあります。
```
訳注をする場合は訳注を明記してください。（現状だと訳文だけで筆者が記述したものかどうかの判断がつきません。）

ymotongpoo reviewed (changes requested):
翻訳そのものよりも、編集に関するコメントです。

tkyshm committed:
fix typo

tkyshm committed:
訳文だけで筆者が記述したものかどうかの判断がつかないので訳注と明記

'訳注:'か'訳注: 'の二通りの書き方が他の箇所でされていたが、とりあえず先にでてきた'訳注:'の形式に揃えた.

shino commented:
@ymotongpoo これは approve 待ちかも?
